src/lambertw.jl

```diff
@@ -0,0 +1,358 @@
+import Base: convert
+
+using Compat
+import Compat.MathConstants  # For clarity, we use MathConstants.e for Euler's number
+
+#### Lambert W function ####
+
+# Use Halley's root-finding method to find
+# x = lambertw(z) with initial point x.
+function _lambertw(z::T, x::T, maxits) where T <: Number
+    two_t = convert(T,2)
+    lastx = x
+    lastdiff = zero(T)
+    converged::Bool = false
+    for i in 1:maxits
+        ex = exp(x)
+        xexz = x * ex - z
+        x1 = x + 1
+        x -= xexz / (ex * x1 - (x + two_t) * xexz / (two_t * x1 ) )
+        xdiff = abs(lastx - x)
+        if xdiff <= 3*eps(abs(lastx)) || lastdiff == xdiff  # second condition catches two-value cycle
+            converged = true
+            break
+        end
+        lastx = x
+        lastdiff = xdiff
+    end
+    converged || warn("lambertw with z=", z, " did not converge in ", maxits, " iterations.")
+    return x
+end
+
+### Real z ###
+
+# Real x, k = 0
+# This appears to be inferrable with T=Float64 and T=BigFloat, including if x=Inf.
+# The fancy initial condition selection does not seem to help speed, but we leave it for now.
+function lambertwk0(x::T, maxits)::T where T<:AbstractFloat
+    isnan(x) && return(NaN)
+    x == Inf && return Inf # appears to return convert(BigFloat,Inf) for x == BigFloat(Inf)
+    one_t = one(T)
+    oneoe = -one_t/convert(T,MathConstants.e)  # The branch point
+    x == oneoe && return -one_t
+    oneoe <= x || throw(DomainError(x))
+    itwo_t = 1/convert(T,2)
+    if x > one_t
+        lx = log(x)
+        llx = log(lx)
+        x1 = lx - llx - log(one_t - llx/lx) * itwo_t
```

Comment:
In generic implementations one should avoid `Float64` literals since they can cause undesired promotions:
```suggestion
        x1 = lx - llx - log1p(- llx/lx) / 2
```